tests/shardddl3/run.sh

```diff
@@ -871,6 +871,1040 @@ function DM_112() {
+    run_sql_source2 "insert into ${shardddl1}.${tb2} values(9,9);"
+
+    run_sql_source2 "alter table ${shardddl1}.${tb2} drop index uk;"
+    run_sql_source1 "insert into ${shardddl1}.${tb1} values(10,10);"
+    run_sql_source2 "insert into ${shardddl1}.${tb1} values(11,11);"
+    run_sql_source2 "insert into ${shardddl1}.${tb2} values(12,12);"
+
+    # FIXME: DM should report an error to user that this operation may not be able to rollback.
+    check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
+}
+
+# Drop the unique key.
+function DM_139 {
+    run_case 139 "double-source-pessimistic" \
+        "run_sql_source1 \"create table ${shardddl1}.${tb1} (id int primary key, a int, unique key uk(a));\"; \
+         run_sql_source2 \"create table ${shardddl1}.${tb1} (id int primary key, a int, unique key uk(a));\"; \
+         run_sql_source2 \"create table ${shardddl1}.${tb2} (id int primary key, a int, unique key uk(a));\"" \
+        "clean_table" "pessimistic"
+
+    run_case 139 "double-source-optimistic" \
+        "run_sql_source1 \"create table ${shardddl1}.${tb1} (id int primary key, a int, unique key uk(a));\"; \
+         run_sql_source2 \"create table ${shardddl1}.${tb1} (id int primary key, a int, unique key uk(a));\"; \
+         run_sql_source2 \"create table ${shardddl1}.${tb2} (id int primary key, a int, unique key uk(a));\"" \
+        "clean_table" "optimistic"
+}
+
+function DM_Partition_CASE {
+    run_sql_source1 "insert into ${shardddl1}.${tb1} values(100),(101),(102),(103),(104),(105);"
+    run_sql_source2 "insert into ${shardddl1}.${tb1} values(200),(201),(202),(203),(204),(205);"
+    run_sql_source2 "insert into ${shardddl1}.${tb2} values(300),(301),(302),(303),(304),(305);"
+
+    # Add partitioning.
+    run_sql_source1 "alter table ${shardddl1}.${tb1} partition by range(id)(partition p0 values less than (106));"
+    run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "query-status test" \
+        "ALTER TABLE \`${shardddl1}\`.\`${tb1}\` PARTITION BY RANGE (\`id\`) (PARTITION \`p0\` VALUES LESS THAN (106))" 1 \
+        "alter table partition is unsupported" 1
+    run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" "handle-error test skip"
+
+    # Add new partition.
+    run_sql_source1 "alter table ${shardddl1}.${tb1} add partition (partition p1 values less than (112));"
+    run_sql_source1 "insert into ${shardddl1}.${tb1} values(106),(107),(108),(109),(110),(111);"
+    run_sql_source2 "insert into ${shardddl1}.${tb1} values(206),(207),(208),(209),(210),(211);"
+    run_sql_source2 "insert into ${shardddl1}.${tb2} values(306),(307),(308),(309),(310),(311);"
+    run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "query-status test" \
+        "ALTER TABLE \`${shardddl1}\`.\`${tb1}\` ADD PARTITION (PARTITION \`p1\` VALUES LESS THAN (112))" 1 \
+        "Partition management on a not partitioned table is not possible" 1
```

Comment:
the error message here seems caused by above "alter table partition is unsupported". maybe we should test if a partition table could be synced at beginning, and if it can, test following add partition DDL could succeed

Reply:
I mistakenly thought  that TiDB doesn't support partition😂. I have re-implemented the partition related test cases.